[BB-1014] Implement a docker based devstack

singuliere opened the pull request:
- Run a shell in the container, with the current directory mounted inside it:

    $ make devstack.shell
    singuliere@2fa899e089b8:/opt/opencraft$ make test.unit

- Run all tests, using the current directory:

    $ make devstack.shell -- make test

- Run Ocim using the current directory and connect to it

    $ make devstack.shell -- make migrate
    $ make devstack.shell -- make manage createsuperuser
    $ make devstack.shell -- make run.dev
    $ firefox http://localhost:5000/

dupainaulevain commented:
For the record @xitij2000 suggested to cherry-pick the good bits from a [similar work he authored some time ago](https://github.com/open-craft/opencraft/tree/kshitij/docker_support)

singuliere commented:
> For the record @xitij2000 suggested to cherry-pick the good bits from a [similar work he authored some time ago](https://github.com/open-craft/opencraft/tree/kshitij/docker_support)

The **yes "no"** is interesting. I did not know `yes` had an argument!

antoviaque commented:
@singuliere Thanks for the contribution : )

@dupainaulevain @UmanShahzad Do we already have a review scheduled for this? If not, would it go in the upcoming sprint?

dupainaulevain commented:
@dupainaulevain there is nothing even remotely scheduled on that topic AFAIK :-)

UmanShahzad commented:
@antoviaque cc @dupainaulevain @singuliere https://tasks.opencraft.com/browse/BB-1014

antoviaque commented:
@UmanShahzad Thanks for creating the task! :+1:

giovannicimolin commented on devstack/Makefile:
```diff
@@ -0,0 +1,34 @@
+DEVSTACK_COMPOSE_FILES=-f devstack/devstack-compose-ocim.yml.tmp
+
+ifeq ($(firstword $(MAKECMDGOALS)),$(filter $(firstword $(MAKECMDGOALS)),devstack.shell))
```
These lines are a little hard to read and know what they are doing.

singuliere replied:
I could not agree more :-) I'll remove them.

giovannicimolin commented on devstack/Makefile:
```diff
@@ -0,0 +1,34 @@
+
+DEVSTACK_BUILD=docker-compose ${DEVSTACK_COMPOSE_FILES} build
+
+devstack.build: devstack/devstack-compose-ocim.yml.tmp  ## Rebuild the devstack docker image
```
A provisioning command can replace this build method.

singuliere replied:
there will be a need for a `docker build` even if it is followed by an ansible run because the xenial image does not come with a ssh server or an unprivileged user that has sudo rights.

singuliere commented on Makefile:
```diff
@@ -25,8 +25,10 @@ COVERAGE_THRESHOLD ?= 94
 WORKERS ?= 3
 WORKERS_LOW_PRIORITY ?= 3
 SHELL ?= /bin/bash
-HONCHO_MANAGE := honcho run python3 manage.py
```
ok

singuliere commented on README.md:
```diff
@@ -126,6 +126,25 @@ Vagrant will map port 5000 inside the virtual machine to port 5000 on the host.
 
+### Docker devstack installation and usage
+
+- [Install docker](https://docs.docker.com/install/)
```
there is no need for docker-compose

singuliere replied:
what was I thinking... there is a need for docker-compose indeed, good catch!

singuliere commented on Makefile:
```diff
@@ -150,6 +152,8 @@ test: clean test.quality test.unit test.migrations_missing test.js test.browser
 test.one: clean
 	$(HONCHO_MANAGE_TESTS) test $(RUN_ARGS)
 
+include devstack/Makefile
```
ok

singuliere commented:
> Can we simplify the commands needed for running the devstack? I think that a nice improvement would be to have commands like these: 

That's all good ideas ! I'll do that :-)

>  Why not use the playbooks that are currently being used to provision the Vagrant devstack to provision Ocim inside the container? The devstack/devstack.dockerfile is basically remaking the steps already done and maintained on our ansible playbooks.

I wanted to do that at first, however it turned out to not help as much as I thought it would, but maybe I'm missing something. Here is what I see:

From [Vagrantfile](https://github.com/open-craft/opencraft/blob/master/Vagrantfile#L23) I understand [ocim-vagrant.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/ocim-vagrant.yml) is called, sets some variables and runs tasks from [common.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/common.yml) and  [vagrant.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/vagrant.yml). When compared with the dockerfile, the following are duplicated:

1. 100% duplicated **make install_system_dependencies** at [this line of common.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/common.yml#L4) 
1. setup the virtualenv but also ensure it is updated every time the requirement files changes **python3 -m venv venv && . venv/bin/activate...** at [this line of common.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/common.yml#L8)
1. install database clients & servers 100% duplicated **DEBIAN_FRONTEND=noninteractive apt-get install -y...** [this line of vagrant.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/vagrant.yml#L14)
1. postgres version 10 instead of version 9 **echo 'deb http://apt.postgresql...** [this line of vagrant.yml](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/ocim/tasks/vagrant.yml#L26)

In order to be ready to run **make test.unit test.js test.browser**  the dockerfile also prepares the image using code borrowed from circle.yml. My main goal was to be able to repeat and debug errors reported by CircleCI.

1. [install and run consul](https://github.com/open-craft/opencraft/blob/master/circle.yml#L63)
1. [install xvfb and a compatible firefox](https://github.com/open-craft/opencraft/blob/master/circle.yml#L5)

The rest of the dockerfile (about half of the lines) is docker specific.

So, if I did not miss anything, running ansible to provision the dockerfile would require to:

* create ocim-docker.yml on the model of ocim-vagrant.yml with appropriate variables
* create a list of tasks in docker.yml
* factor out of vagrant.yml the tasks that are common to both vagrant & docker (i.e. the virtualenv is done differently for a reason) into a devstack.yml
* create a docker.yml list of tasks including devstack.yml 

And the benefits would be to avoid duplicating:

* make install_system_dependencies
* make install_system_db_dependencies

What should be done, definitely, is make it so the dockerfile runs `make install_system_db_dependencies` instead of list dependencies, otherwise that would be very hard to maintain. 

Another thing would be to decide if postgres 9 or postgres 10 should be used and move the setup out of Ansible so it can be shared between vagrant and circle.yml (and docker).

Once that's done, the only remaining difference is how the virtualenv is set. This should however not cause any maintenance problem since the dependencies are read from the same files by circle/vagrant/docker. 

What do you think?

singuliere commented:
@giovannicimolin I think I've addressed all the comments, except for the ongoing discussion about how to approach the provisioning so it is not a maintenance burden. And split out the other commit to https://github.com/open-craft/opencraft/pull/406 as requested.

I really appreciate your review and your insights: it's feels very good that this volunteer contribution gets attention. Thanks :+1: :sparkles:

giovannicimolin commented on devstack/devstack.dockerfile:
```diff
@@ -0,0 +1,68 @@
+#
+# Install dependencies (this should all be in debian_packages.lst but not sure yet why it's not already)
+#
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y git sudo python python3-dev python3 python3-venv ca-certificates unzip openssh-server mongodb mysql-server mysql-client unzip gcc make
```
Why not start from image `python:3.5` instead of installing it?

Additionally, instead of installing all the components needed on a single container, why not configure standalone `postgres:10.7`, `mongo` and `consul` containers and link them with the Ocim container using `docker-compose`? (similarly to how it's done with the Open edX devstack)

singuliere replied:
It's a matter of convenience and is debatable indeed! My reasons for preferring to start from a base Xenial container instead of python:3.5 is that I know exactly what this image contains. I would have to investigate what python:3.5 has or what it is missing. Is it based on Xenial? etc.

Regarding the services, I have a strong preference for running them in a all-in-one container for development convenience rather than assembling them with docker-compose.yml, because:

* It is what Vagrant does currently
* It is easier to debug, restart, reconfigure a depedency running on the same machine rather than in another container
* Developers would have to learn a different way of interacting with servers compared to what they are currently used to (i.e. all of them running in the same Vagrant VM)

giovannicimolin replied:
@singuliere Regarding your comments:

> Regarding the services, I have a strong preference for running them in a all-in-one container for development convenience rather than assembling them with docker-compose.yml

If this is the main goal, using docker will only make it harder, because of the mount points and permissions, which Vagrant already takes care for us.
The goal of having a docker devstack with multiple containers is to have each service of the system running independently and without interfering with each other's packages and dependencies, and putting everything on the same container makes Docker basically the same as the current Vagrant devstack + tech debt.

> It is easier to debug, restart, reconfigure a dependency running on the same machine rather than in another container

This point is debatable, since the goal of having different containers is to make the services independent and with less risk of updating one component and breaking others.

> Developers would have to learn a different way of interacting with servers compared to what they are currently used to (i.e. all of them running in the same Vagrant VM)

I agree with this, but we already work mainly on Open edX using the docker devstack, so it wouldn't be that much of news.

My point is that it is only worth switching from our Vagrant devstack to a Docker devstack if:
- The docker based devstack improves or eases our current workflow
- Doesn't add a significant amount of tech debt
- Completely replaces our Vagrant devstack

Another point that should be raised is that currently, we use the same playbooks to create the devstack that we use to deploy Ocim (changing some roles, of course), so we also have the plus side of testing our current deployment strategies on the Vagrant machine.

To move this forward, I will have to start an internal discussion with the team to see if switching to a Docker based devstack is in the interest of the project and if we have the budget to get this moving forward.

Thanks for your patience and your contribution! I'll get back to you as soon as I have some news.

giovannicimolin commented on devstack/devstack.dockerfile:
```diff
@@ -0,0 +1,68 @@
+#
+COPY requirements.txt .
+COPY cleanup_utils/requirements.txt cleanup_utils/requirements.txt
+RUN python3 -m venv venv && . venv/bin/activate && pip install --upgrade pip && pip install --upgrade virtualenv && pip install -r requirements.txt && pip install -r cleanup_utils/requirements.txt && chown -R $USER_NAME venv
```
Since we're on a isolated container running a single application, we don't need a virtualenv, specially when building container directly from `python` builds.

singuliere replied:
IIRC there are (very) subtle differences when running outside of a virtualenv. This is only a vague recollection. If I'm not mistaken Ocim runs from a virtualenv in production and that circle.yml also installs a virtualenv for testing purposes. If that's not a blocker, I'd rather not try to verify if running outside of a virtualenv makes a difference in the context of this pull request. Maybe it's just me being over-cautious and not trying something that nobody does ;-)

giovannicimolin commented on devstack/devstack.dockerfile:
```diff
@@ -0,0 +1,68 @@
+           -e "s/%%USER_ID%%/$USER_ID/g" \
+	   /usr/local/bin/setup-devstack.sh
+COPY devstack/setup-devstack.service /etc/systemd/system/setup-devstack.service
+RUN systemctl enable setup-devstack
```
If this is a one-shot command, why not run it while provisioning the devstack? 
Eg: When running `make devstack.provision`, create dbs, users, apply migrations and leave the server ready to run.

singuliere replied:
This goes back to what I think is a misunderstanding on the scope of this PR: it does not deploy Ocim.

giovannicimolin commented on devstack/devstack.dockerfile:
```diff
@@ -0,0 +1,68 @@
+# This is 100% duplicated from .env.test although only a few values needs replacement, it should be split
+#
+ENV HONCHO_TEST_ENV=devstack/honcho.env.test.tmp
+ENV HONCHO_ENV=devstack/honcho.env.tmp
```
How is the Ocim server run in the container?
I think we need to add a `ENTRYPOINT` parameter, that way, when the container is started, Ocim is automatically run and the output can be viewed by either `attach` or `log` docker commands.

singuliere replied:
> How is the Ocim server run in the container?

It does not. It could if the user wanted to but that's their call. The primary purpose is development, therefore running `make test.unit` etc.

> I think we need to add a ENTRYPOINT parameter, that way, when the container is started, ...

Note that the container runs in the background and the entrypoint must be `systemd`.

giovannicimolin replied:
> The primary purpose is development, therefore running make test.unit etc.

The main goal of our current devstack is to provide a development environment that is easy to use and provides a fully working Ocim instance and also runs the unit tests. 

This is required because when developing on Ocim, we need to **not only run the unit tests locally**, but also check UI changes, deployment workflow and use the Django admin to set variables and verify our work. 

> Note that the container runs in the background and the entrypoint must be systemd.
Not necessarily, the entrypoint can also be a `supervisor` monitoring and restarting the django process automatically.

singuliere commented:
> The problem with this approach is that if we change anything related to the deployment of Ocim or the dependencies, we'll have to update two places (playbooks and dockerfile) ...

I think there is a misunderstanding on the scope of this pull request. It does **not** deploy Ocim. It only installs the dependencies and it does so using makefiles targets, in the same way circle.yml does. The installation of dependencies already need to be updated in two places (Vagrant dedicated playbook and circle.yml) and would need to be updated in a third (dockerfile). But this update has a very low variance and I'm guessing that's why it was kept as is so far in the Vagrant dedicated playbook and circle.yml.

singuliere commented on Makefile:
```diff
@@ -32,7 +32,7 @@ RUN_JS_TESTS := xvfb-run --auto-servernum jasmine-ci --logs --browser firefox
 
 # For `test.one` use the rest as arguments and turn them into do-nothing targets
-ifeq ($(firstword $(MAKECMDGOALS)),$(filter $(firstword $(MAKECMDGOALS)),test.one manage))
+ifeq ($(firstword $(MAKECMDGOALS)),$(filter $(firstword $(MAKECMDGOALS)),test.one manage devstack.shell))
```
It is but I did not write it so I don't feel too guilty about its complexity ;-) It allows the following:

    make manage a b c

which calls

    honcho -e ${HONCHO_ENV} run python3 manage.py a b c

singuliere commented:
The origin of the confusion is likely to be my (bad) idea to call it `devstack` because in the context of Open edX, `devstack` deploys and runs the lms/cms! It is closer to what `make shell does in the `edx-platform` repository.

singuliere commented:
I reorganized the dockerfile for clarity. And make it so it relies on `make install_system_db_dependencies` to avoid duplication.

singuliere commented:
I understand your concerns and why this docker based devstack would create more problems than it resolves. Thanks for taking the time to review it and explain the details. I never actually ran Ocim in a test environment, reason why I did not understand the bigger picture.

UmanShahzad commented on devstack/devstack.dockerfile:
```diff
@@ -0,0 +1,80 @@
+# Install consul
+#
+RUN apt-get install -y unzip && wget -P /tmp https://releases.hashicorp.com/consul/1.2.1/consul_1.2.1_linux_amd64.zip && unzip /tmp/consul_1.2.1_linux_amd64.zip -d /usr/local/bin
+COPY devstack/consul.service /etc/systemd/system/consul.service
```
The approach here should be to start Consul as a separate container within the same network as Ocim through docker-compose, which manages this easily. Same with postgres and basically everything else.
